hoomd/box.py

```diff
@@ -0,0 +1,402 @@
+import numpy as np
+from functools import partial
+import hoomd._hoomd as _hoomd
+
+
+def _make_vec_three(vec, vec_factory, scalar_type):
+    """A helper function for converting Python sequences to HOOMD T3 classes.
+
+    Args:
+        vec (Sequence[X] or X): A sequence of type X where X is the type
+            returned by scalar_type.
+        vec_factory (function): A function from `hoomd._hoomd` that makes a T3
+            class (e.g. Scalar3, Int3).
+        scalar_type (function): A function to convert a Python value into the
+            base type for the vec_factory function. For `Scalar3` this would be
+            `float`.
+    """
+    try:
+        l_vec = len(vec)
+    except TypeError:
+        try:
+            v = scalar_type(vec)
+        except (ValueError, TypeError):
+            raise ValueError("Expected value of type {}.".format(scalar_type))
+        else:
+            return vec_factory(v, v, v)
+    if l_vec == 3:
+        try:
+            return vec_factory(scalar_type(vec[0]),
+                               scalar_type(vec[1]),
+                               scalar_type(vec[2]))
+        except (ValueError, TypeError):
+            raise ValueError("Expected values of type {}.".format(scalar_type))
+    else:
+        raise ValueError("Expected three or one value received {}.".format(
+            len(vec)))
+
+
+_make_scalar3 = partial(_make_vec_three,
+                        vec_factory=_hoomd.make_scalar3,
+                        scalar_type=float)
+
+_make_int3 = partial(_make_vec_three,
+                     vec_factory=_hoomd.make_int3,
+                     scalar_type=int)
+
+_make_char3 = partial(_make_vec_three,
+                      vec_factory=_hoomd.make_char3,
+                      scalar_type=int)
+
+
+def _to_three_array(vec, dtype=None):
+    return np.array((vec.x, vec.y, vec.z), dtype=dtype)
+
+
+class Box:
+    R""" Define box dimensions.
+
+    Args:
+        Lx (float): box extent in the x direction (distance units).
+        Ly (float): box extent in the y direction (distance units).
+        Lz (float): box extent in the z direction (distance units).
+        xy (float): tilt factor xy (dimensionless).
+        xz (float): tilt factor xz (dimensionless).
+        yz (float): tilt factor yz (dimensionless).
+
+    Simulation boxes in hoomd are specified by six parameters, ``Lx``, ``Ly``,
+    ``Lz``, ``xy``, ``xz``, and ``yz``. `Box` provides a way to specify all
+    six parameters for a given box and perform some common operations with them.
+    A `Box` can be passed to an initialization method or to assigned to a
+    saved :py:class:`State` variable (``state.box = new_box``) to set the
+    simulation box.
+
+    Access attributes directly::
+
+        b = hoomd.box.Box(L=20)
+        b.xy = 1.0
+        b.yz = 0.5
+        b.Lz = 40
+
+    .. rubric:: Two dimensional systems
+
+    2D simulations in HOOMD use boxes with ``Lz == 0``. 2D boxes ignore ``xz``
```

Comment:
Is `Lz == 0` for 2D simulations a change from the current version, and if so, is there a reason to change?

Reply:
This is a change from v2.x. The primary reason is that this makes more intuitive sense. When thinking of a 2D box the length of a third dimension should be 0. Using the z axis is just following convention. The previous behavior of setting it just through a flag and requiring a non-zero length z-axis is a hold over from when that would cause a divide by zero error.

Reply:
It indeed makes sense that the 3rd dimension should be 0 for a 2D system. I'm still a little uneasy checking for dimensionality by looking at the value of Lz though. That should be a constant parameter of your simulation that always returns the expected value (and not if something, however unphysical it may be) causes your Lz to go to 0.

Reply:
Part of the reason for this change is so that Box knows whether it is 2D or 3D. If the dimensionality is a separate property of the state, then just having a Box object is not enough to know whether to compute the volume or the area (for example). Using Lz==0 seems the most reasonable choice without adding additional fields to `BoxDim`.

What change to the box do you think would be able to force Lz to precisely zero?

Reply:
I actually think this change will make it easier to catch bugs. If a 3D NPT simulation suddenly has Lz=0, I'd expect it to break due to a divide by zero at some point anyway, but if not any usage of dimensions would also raise a red flag if it changed.